src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -497,9 +499,27 @@ public override string MetadataName
         }
 
         /// <summary>
-        /// If this type is a file-local type, returns the syntax tree where this type is visible. Otherwise, returns null.
+        /// If this type is a file-local type, returns an identifier for the file this type was declared in. Otherwise, returns null.
         /// </summary>
-        internal abstract SyntaxTree? AssociatedSyntaxTree { get; }
+        internal abstract FileIdentifier? AssociatedFileIdentifier { get; }
+
+        internal struct FileIdentifier
+        {
+            public ImmutableArray<byte> FilePathChecksum { get; init; }
+            public string DisplayFilePath { get; init; }
+
+            public static FileIdentifier FromSyntaxTree(SyntaxTree tree)
+            {
+                using var sha256 = SHA256.Create();
+                var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(tree.FilePath));
```

Comment:
>tree.FilePath [](http://example.com/codeflow?start=69&length=13)

Is this the full path or the path relative to the project root?

Reply:
Strictly speaking: anyone who references MS.CA.CS can create SyntaxTrees in memory and pass whatever string they want for the file path.

In practice: this is the full path after path mapping. That means all our existing mechanisms for ensuring determinism in presence of 'CallerFilePath' attributes, file paths in PDBs, etc. are in place here.